File: 2015/day2/js/2015.2.mjs
import * as readline from "node:readline";
import { createReadStream } from "node:fs";

const rl = readline.createInterface({
    input: createReadStream("../input.txt"),
    crlfDelay: Infinity, // crlfDelay option to recognize all instances of CR LF ('\r\n') in input.txt as a single line break.
});

let totalPaper = 0;
let totalRibbon = 0;
let separator = "x";
let dimensions = [];

rl.on("line", (line) => {
    let [l, w, h] = line.split(separator).map(value => +value);
    dimensions = [l, w, h].sort((a, b) => a - b);

    totalPaper += (2 * l * w) + (2 * w * h) + (2 * h * l);
    totalPaper += dimensions[0] * dimensions[1];

    totalRibbon += dimensions[0] + dimensions[0] + dimensions[1] + dimensions[1];
    totalRibbon += l * w * h;
});

rl.on("close", () => {
    console.log(`Total paper needed : ${totalPaper} square feet`);
    console.log(`Total ribbon needed : ${totalRibbon} square feet`);
});

File: 2015/day2/c++/2015.2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include <algorithm>

using namespace std;

struct dimensions {
	int l;
	int w;
	int h;
};

dimensions getDimensions(string input) {
	string delimiter = "x";
	int dimensionvalues[3];
	int start = 0;
	int end = 0;
	int iteration = 0;

	while (iteration < 3) {
		end = input.find(delimiter, start);
		dimensionvalues[iteration] = stoi(input.substr(start, end - start));
		start = end + delimiter.size();
		iteration++;
	}

	dimensions d;
	d.l = dimensionvalues[0];
	d.w = dimensionvalues[1];
	d.h = dimensionvalues[2];

	return d;
}

int main()
{
	ifstream inputFile("input.txt");

	string line;
	int totalPaper = 0;
	int totalRibbon = 0;
	dimensions d;
	array<int, 3>dimensions;

	while (getline(inputFile, line)) {
		d = getDimensions(line);
		dimensions = {d.l, d.w, d.h};
		
		std::sort(dimensions.begin(), dimensions.end());

		totalPaper += (2 * d.l * d.w) + (2 * d.w * d.h) + (2 * d.h * d.l);
		totalPaper += dimensions[0] * dimensions[1];

		totalRibbon += dimensions[0] + dimensions[0] + dimensions[1] + dimensions[1];
		totalRibbon += d.l * d.w * d.h;

		//std::cout << "l :" << d.l << ", w :" << d.w << ", h :" << d.h << "\n";
	}

	inputFile.close();

	std::cout << "total paper needed : " << totalPaper << " square feet";
	std::cout << "\n";
	std::cout << "total ribbon needed : " << totalRibbon << " square feet";
}

